File: libs/document-processor/src/types/result.ts
export interface Ok<T> {
  readonly ok: true;
  readonly value: T;
}

export interface Err<E> {
  readonly ok: false;
  readonly error: E;
}

export type Result<T, E> = Ok<T> | Err<E>;

export function ok<T>(value: T): Ok<T> {
  return { ok: true, value };
}

export function err<E>(error: E): Err<E> {
  return { ok: false, error };
}

export function isOk<T, E>(result: Result<T, E>): result is Ok<T> {
  return result.ok;
}

export function isErr<T, E>(result: Result<T, E>): result is Err<E> {
  return !result.ok;
}

export function map<T, E, U>(
  result: Result<T, E>,
  mapper: (value: T) => U,
): Result<U, E> {
  return result.ok ? ok(mapper(result.value)) : result;
}

export function mapErr<T, E, F>(
  result: Result<T, E>,
  mapper: (error: E) => F,
): Result<T, F> {
  return result.ok ? result : err(mapper(result.error));
}

export function andThen<T, E, U>(
  result: Result<T, E>,
  mapper: (value: T) => Result<U, E>,
): Result<U, E> {
  return result.ok ? mapper(result.value) : result;
}

export function unwrapOr<T, E>(result: Result<T, E>, fallback: T): T {
  return result.ok ? result.value : fallback;
}

export function unwrapOrElse<T, E>(
  result: Result<T, E>,
  getFallback: (error: E) => T,
): T {
  return result.ok ? result.value : getFallback(result.error);
}

export function match<T, E, U>(
  result: Result<T, E>,
  handlers: { readonly ok: (value: T) => U; readonly err: (error: E) => U },
): U {
  return result.ok ? handlers.ok(result.value) : handlers.err(result.error);
}

File: references/java-sdk/src/main/java/blue/language/processor/types/Result.java
package blue.language.processor.types;

import java.util.Objects;
import java.util.function.Function;

/**
 * Immutable functional result type mirroring JS Result helpers.
 */
public final class Result<T, E> {
    private final boolean ok;
    private final T value;
    private final E error;

    private Result(boolean ok, T value, E error) {
        this.ok = ok;
        this.value = value;
        this.error = error;
    }

    public static <T, E> Result<T, E> ok(T value) {
        return new Result<T, E>(true, value, null);
    }

    public static <T, E> Result<T, E> err(E error) {
        return new Result<T, E>(false, null, error);
    }

    public boolean isOk() {
        return ok;
    }

    public boolean isErr() {
        return !ok;
    }

    public T value() {
        if (!ok) {
            throw new IllegalStateException("Result is Err");
        }
        return value;
    }

    public E error() {
        if (ok) {
            throw new IllegalStateException("Result is Ok");
        }
        return error;
    }

    public <U> Result<U, E> map(Function<? super T, ? extends U> mapper) {
        Objects.requireNonNull(mapper, "mapper");
        if (ok) {
            return Result.ok(mapper.apply(value));
        }
        @SuppressWarnings("unchecked")
        Result<U, E> same = (Result<U, E>) this;
        return same;
    }

    public <F> Result<T, F> mapErr(Function<? super E, ? extends F> mapper) {
        Objects.requireNonNull(mapper, "mapper");
        if (ok) {
            @SuppressWarnings("unchecked")
            Result<T, F> same = (Result<T, F>) this;
            return same;
        }
        return Result.err(mapper.apply(error));
    }

    public <U> Result<U, E> andThen(Function<? super T, Result<U, E>> mapper) {
        Objects.requireNonNull(mapper, "mapper");
        if (ok) {
            return mapper.apply(value);
        }
        @SuppressWarnings("unchecked")
        Result<U, E> same = (Result<U, E>) this;
        return same;
    }

    public T unwrapOr(T fallback) {
        return ok ? value : fallback;
    }

    public T unwrapOrElse(Function<? super E, ? extends T> getFallback) {
        Objects.requireNonNull(getFallback, "getFallback");
        return ok ? value : getFallback.apply(error);
    }

    public <U> U match(Function<? super T, ? extends U> okHandler,
                       Function<? super E, ? extends U> errHandler) {
        Objects.requireNonNull(okHandler, "okHandler");
        Objects.requireNonNull(errHandler, "errHandler");
        return ok ? okHandler.apply(value) : errHandler.apply(error);
    }
}
